### ISSUE TITLE
Feature question - data driven testing

### ISSUE THREAD

**Issue by QA1MTG** [SeleniumHQ/selenium-ide#183]
Is there ability in the new IDE to load data from excel spreadsheets and use the values in the test?

**Comment by jesg** [comment]
i'm also interested in this feature.  i'd be fine with the data being csv or json.

@corevo i can help build this feature.

**Comment by QA1MTG** [comment]
Currently I have a mechanism that was built in the user-extension.js to do data driven based on loading an XML file.  Since user-extension isn't available, if there's an easier way to do via Excel/CSV/Json that would be really helpful.

**Comment by corevo** [comment]
This is quite challenging, the main issue is, that browser vendors (i.e. Mozilla & Google) don't let extensions access your filesystem anymore, meaning I can't read or write files unless you open them through dialog, or download them.  
The problem is that if you're using a csv file, it would be a bit frustrating, having to open a dialog and navigate to it every time you open the test case.  
Also since we don't have filesystem we can't reliably fetch it when it is ran from command line.  
  
We'll have to think of a clever, and good way (in terms of ux) to do it, atm I don't have a good solution for it.

**Comment by QA1MTG** [comment]
from a UX perspective, what if you added a tab next to log and had the file with the data display?  maybe it's a plugin that allows you to view excel files/csv/xml?

would it be easier to have the file type be a google spreadsheet full of data?  this way it's not an external file?

I could see issues if it's data that needs to be confidential

**Comment by jesg** [comment]
i'd be fine with uploading the data once and not being able to modify the data.  or i'd be fine with fetching the data from a url though that seems more complicated.

**Comment by treg** [comment]
How about reading data, storing screenshots from Dropbox or similar?

-OR-

Just add url handler to website to get/put data?

**Comment by jesg** [comment]
@corevo shouldn't we be able to store large blobs with the [file system api (non-standard)](https://developer.mozilla.org/en-US/docs/Web/API/File_and_Directory_Entries_API) which provides a sandboxed filesystem.  or we could use [indexedDB](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API) which is standard.

where does the project state get stored?  i was not able to figure this out though it looks like reactjs reacts when the mutation property is modified.

**Comment by corevo** [comment]
@QA1MTG I don't like that idea, it will make the IDE a spreadsheet editor as well, which will never be as good as any other product on the market today.  

@treg I thought back in the day about downloading through static websites, but that would put your data at risk, as it will be public accessible, unless you host it on localhost, or we provide integration with different services.

@jesg Uploading once is a bad idea, because soon you'll forget what data was there.
fs api, is only partially available on ff, but in any case if we use either fs api or indexDB, the test case will not be portable from machine to machine.  
At the end all the data needs to be stored in the `side` file, and I really don't want to add any more attachments to it, as it'll make it hard to maintain.

**Comment by QA1MTG** [comment]
@corevo I see your point, although I was only thinking of the spreadsheet viewer to be read only data, and not an editor.  Just having the ability to see the data that the test is using.  Any manual updates would be in the spreadsheet and loaded again if changes are necessary.

**Comment by jesg** [comment]
@corevo i disagree with your concerns about @treg's proposal.  most corporate organizations are on a vpn so this isn't as big of an issue as you are making it.

**Comment by corevo** [comment]
@jesg even if you're using a vpn, the IDE isn't built for corporate organizations alone.  
We needs to provide a solution that would keep the data safe, both for organizations and individuals.  
Don't get me wrong, I want this feature to be implemented, but it needs to be done in a right way.

**Comment by treg** [comment]
Since you are building a plug-in module,
Could the data be loaded from there?

**Comment by corevo** [comment]
Not exactly, any plugin built is subject to the same limitation any other extension has.  
Meaning if you plan on doing that using a plugin, that plugin will face the same difficulties the IDE has.

**Comment by treg** [comment]
I guess what I was thinking is to build a plug-in with hardcoded test data
and using that within the IDE..

**Comment by corevo** [comment]
We could make it work, it's not part of the current plugin system, and we have to find a way to give plugins the correct API calls to enable that.  
Although I would like to find a solution that would work natively in the IDE.

**Comment by testanalyst** [comment]
@tourdedave Is this enhancement prioritized yet for design and development?

**Comment by tourdedave** [comment]
No, but I started working on it in my spare time to demonstrate how plugins could enable this.

I have a PR that needs some work to add functionality in control flow and the plugins API ([link](https://github.com/SeleniumHQ/selenium-ide/pull/584)) and I have a plugin that can be used to upload a CSV and pass it to the IDE ([link](https://github.com/tourdedave/selenium-ide-plugins)). You can see a demo of it in action [here](https://www.dropbox.com/s/e782148j2pq44uy/side-plugins-data-driven.mp4?dl=0).

As I mentioned, there's still work to be done. Not sure when it will land yet. I'll post here when something is available.

**Comment by myronrezzo** [comment]
> No, but I started working on it in my spare time to demonstrate how plugins could enable this.
> 
> I have a PR that needs some work to add functionality in control flow and the plugins API ([link](https://github.com/SeleniumHQ/selenium-ide/pull/584)) and I have a plugin that can be used to upload a CSV and pass it to the IDE ([link](https://github.com/tourdedave/selenium-ide-plugins)). You can see a demo of it in action [here](https://www.dropbox.com/s/e782148j2pq44uy/side-plugins-data-driven.mp4?dl=0).
> 
> As I mentioned, there's still work to be done. Not sure when it will land yet. I'll post here when something is available.



@tourdedave, the [video](https://www.dropbox.com/s/e782148j2pq44uy/side-plugins-data-driven.mp4?dl=0) just demonstrates how a plugin could work by providing the data driven capabilites to the Selenium IDE, right? [The repository plugin](https://github.com/tourdedave/selenium-ide-plugins) is not yet functional, correct?

**Comment by tourdedave** [comment]
@myronrezzo It's functional.

I released the plugin changes necessary for it last month. If you upload a CSV file with the file uploader, it will get parsed, its contents sent to Selenium IDE, and stored in a variable called `fileContents`. From there you can use `forEach` to iterate it.

Here's an example project file that demonstrates it: [data-driven.side.zip](https://github.com/SeleniumHQ/selenium-ide/files/3247282/data-driven.side.zip).

**Comment by p2635** [comment]
Hi @tourdedave , would you be able to send me the csv file that goes along with the project as well? Is it literally a column with username and a column for password?

**Comment by x9sim9** [comment]
One thing I would recommend to the dev team is is to take a look at postman and postman interceptor. So postman itself is a chrome app, and it uses the postman interceptor plugin to interact with the browser. If selenium ide was ported to a chrome app it would remove all the limitations that extensions have regarding files and then it could simply have a companion extension which interacts with the browser.

**Comment by corevo** [comment]
Chrome does not accept chrome apps anymore, it has been discontinued.

But as you said postman is a good idea, recently they've ported postman to electron which is precisely what we're doing.

Cheers,
Tomer

**Comment by Fugit185** [comment]
> > No, but I started working on it in my spare time to demonstrate how plugins could enable this.
> > I have a PR that needs some work to add functionality in control flow and the plugins API ([link](https://github.com/SeleniumHQ/selenium-ide/pull/584)) and I have a plugin that can be used to upload a CSV and pass it to the IDE ([link](https://github.com/tourdedave/selenium-ide-plugins)). You can see a demo of it in action [here](https://www.dropbox.com/s/e782148j2pq44uy/side-plugins-data-driven.mp4?dl=0).
> > As I mentioned, there's still work to be done. Not sure when it will land yet. I'll post here when something is available.
> 
> @tourdedave, the [video](https://www.dropbox.com/s/e782148j2pq44uy/side-plugins-data-driven.mp4?dl=0) just demonstrates how a plugin could work by providing the data driven capabilites to the Selenium IDE, right? [The repository plugin](https://github.com/tourdedave/selenium-ide-plugins) is not yet functional, correct?

I have following issue, it writes ${username} to username field and not the username is CSV file?

**Comment by panayot-zhi** [comment]
As a workaround that only works in chrome currently you could enable the extension access to file URLs open the file:// and store the text, although you would have to do the parsing yourself. It is still a way to work around the file access limitations of extensions and kind of parameterize your tests.

**Comment by wolvriene** [comment]
is this resolved? how can we do the same in CLI mode ?

**Comment by Sowndariya** [comment]
> > > No, but I started working on it in my spare time to demonstrate how plugins could enable this.
> > > I have a PR that needs some work to add functionality in control flow and the plugins API ([link](https://github.com/SeleniumHQ/selenium-ide/pull/584)) and I have a plugin that can be used to upload a CSV and pass it to the IDE ([link](https://github.com/tourdedave/selenium-ide-plugins)). You can see a demo of it in action [here](https://www.dropbox.com/s/e782148j2pq44uy/side-plugins-data-driven.mp4?dl=0).
> > > As I mentioned, there's still work to be done. Not sure when it will land yet. I'll post here when something is available.
> > 
> > 
> > @tourdedave, the [video](https://www.dropbox.com/s/e782148j2pq44uy/side-plugins-data-driven.mp4?dl=0) just demonstrates how a plugin could work by providing the data driven capabilites to the Selenium IDE, right? [The repository plugin](https://github.com/tourdedave/selenium-ide-plugins) is not yet functional, correct?
> 
> I have following issue, it writes ${username} to username field and not the username is CSV file?

I am also facing the same issue and loop runs infinitely. Is it fixed?

**Comment by pvnlsainadh** [comment]
this Pl

> No, but I started working on it in my spare time to demonstrate how plugins could enable this.
> 
> I have a PR that needs some work to add functionality in control flow and the plugins API ([link](https://github.com/SeleniumHQ/selenium-ide/pull/584)) and I have a plugin that can be used to upload a CSV and pass it to the IDE ([link](https://github.com/tourdedave/selenium-ide-plugins)). You can see a demo of it in action [here](https://www.dropbox.com/s/e782148j2pq44uy/side-plugins-data-driven.mp4?dl=0).
> 
> As I mentioned, there's still work to be done. Not sure when it will land yet. I'll post here when something is available.

This Workaround have warnings to fix and while i am trying to replicate, we have an open issue , https://github.com/tourdedave/selenium-ide-plugins/issues/1
Due to this, we are unable to proceed

**Comment by matt-o-matic** [comment]
> This is quite challenging, the main issue is, that browser vendors (i.e. Mozilla & Google) don't let extensions access your filesystem anymore, meaning I can't read or write files unless you open them through dialog, or download them.
> The problem is that if you're using a csv file, it would be a bit frustrating, having to open a dialog and navigate to it every time you open the test case.
> Also since we don't have filesystem we can't reliably fetch it when it is ran from command line.
> 
> We'll have to think of a clever, and good way (in terms of ux) to do it, atm I don't have a good solution for it.

I have seen a similar thing done in Word documents that have mail merge enabled.  Essentially when you open the document it detects there is a dependency and if it can't find it right away it prompts you to select the file that satisfies the dependency.  In this case for SIDE it would just always prompt you to select the CSV/JSON/etc file that is needed to make the test work.  No big deal really... and a small price to pay from a UX standpoint for such a valuable feature.

So it would be:
Open project
Parse project file, finds data dependency
Prompt to load dependency file
Repeat until all dependencies are loaded
Profit! :-)

**Comment by corevo** [comment]
Maybe we could, it would be up to the user to decide wether or not he wants to use this feature, and have to load things manually every time.

What do you think we should do about the runner though?

**Comment by matt-o-matic** [comment]
> Maybe we could, it would be up to the user to decide wether or not he wants to use this feature, and have to load things manually every time.
> 
> What do you think we should do about the runner though?

I was thinking about that too... it could be as simple as a command line option to import data files... like --dep-file=/file1.csv --dep-file=/file2.csv etc etc (or I suppose the -c syntax would be a good fit here too)

In the case of it trying to run in the grid... perhaps a runtime compilation of the deps into the json before sending the file to the grid?  Maybe a structure like:

{"Dependencies":[{"filename":"/file1.csv", "data":"[data]"}, {"filename":"/file2.csv", "data":"[data]"}]}

Or even more simple:
{"Dependencies":[{"/file1.csv":"[data]"},{"/file2.csv":"[data]"}]}

So in the code it could be called without needing to loop through the array each time...

Just my thoughts on it after probably not nearly enough time thinking about it... but it would be a super useful feature for sure... and would, in my opinion, make this probably the best no-code testing solution out there for front-end automated QA testing.

Hope this helps. :-)

**Comment by fopaitry** [comment]
I understand that browsers can't access filesystem .
Nethertheless, the SIDE runner is command line and is able to access file system : could you transfert the data with a parameter of the side runner  ?  data stored in memory until the end of run

**Comment by PawelSuwinski** [comment]
With SIDE runner is a just matter of side file preprocessing before running. I did it in some php project as part of a composer  script, I do not have any JS `npm run-script` working example but would use templates concept this way:

1. For CSV data use something like [csv2json](https://www.npmjs.com/package/csvtojson) to get data in JSON format
2. Creating template SIDE file use `store json` with _Target_  like ex. %DATA%
3. In preprocessor replace all '%DATA%' in template side file with target data (ex. using [rexreplace](https://www.npmjs.com/package/rexreplace)) and save it as a target side file (in cache/ tmp area)
4. Run side runner on target side file

2-3 could be also done using some full template engine.
